src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -0,0 +1,278 @@
+
+internal sealed class KestrelMetrics
+{
+    // Note: Dot separated instead of dash.
+    public const string MeterName = "Microsoft.AspNetCore.Server.Kestrel";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentConnectionsCounter;
+    private readonly Histogram<double> _connectionDuration;
+    private readonly Counter<long> _rejectedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedRequestsCounter;
+    private readonly UpDownCounter<long> _currentUpgradedRequestsCounter;
+    private readonly Histogram<double> _tlsHandshakeDuration;
+    private readonly UpDownCounter<long> _currentTlsHandshakesCounter;
+
+    public KestrelMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "current-connections",
+            description: "Number of concurrent connections that are currently active on the server.");
+
+        _connectionDuration = _meter.CreateHistogram<double>(
+            "connection-duration",
+            unit: "s",
+            description: "The duration of connections on the server. Connection duration is measured in seconds.");
+
+        _rejectedConnectionsCounter = _meter.CreateCounter<long>(
+           "rejected-connections",
+            description: "Number of connections rejected by the server. Connections are rejected when the currently active count exceeds the value configured with MaxConcurrentConnections.");
+
+        _queuedConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "queued-connections",
+            description: "Number of connections that are currently queued and are waiting to start.");
+
+        _queuedRequestsCounter = _meter.CreateUpDownCounter<long>(
+           "queued-requests",
+            description: "Number of HTTP requests on multiplexed connections (HTTP/2 and HTTP/3) that are currently queued and are waiting to start.");
+
+        _currentUpgradedRequestsCounter = _meter.CreateUpDownCounter<long>(
+           "current-upgraded-requests",
+            description: "Number of HTTP requests that are currently upgraded (WebSockets).");
+
+        _tlsHandshakeDuration = _meter.CreateHistogram<double>(
+            "tls-handshake-duration",
+            unit: "s",
+            description: "The duration of TLS handshakes on the server.");
+
+        _currentTlsHandshakesCounter = _meter.CreateUpDownCounter<long>(
+           "current-tls-handshakes",
+            description: "Number of TLS handshakes that are currently active on the server.");
+    }
+
+    public void ConnectionStart(BaseConnectionContext connection)
+    {
+        if (_currentConnectionsCounter.Enabled)
+        {
+            ConnectionStartCore(connection);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Why `NoInlining`?

Reply:
`TagList` is a big struct. Trying to isolate the cost of the counter if counters aren't enabled.

Reply:
Can you please elaborate?  What's the cost and how does not inlining mitigate it?

Reply:
The stack space for the big tag list isn't needed if the method is never called. There is probably inlining heuristics that would prevent it from being inlined anyway, I just made it explicit with the attribute.